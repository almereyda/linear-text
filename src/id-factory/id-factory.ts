/** Identity factory. */
export type IDFactory = {
  /** ID of the next. */
  id: number
}

export function IDFactory(factory: IDFactory | undefined = {id: 1}): IDFactory {
  return {id: factory.id}
}

export function makeID(factory: IDFactory): number {
  const id = factory.id
  factory.id++
  return id
}
