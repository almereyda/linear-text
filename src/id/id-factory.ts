import {ID} from './id'

/** Identity factory. */
export type IDFactory = {
  /** ID of the next. */
  id: number
}

export function IDFactory(factory: IDFactory | undefined = {id: 1}): IDFactory {
  return {id: factory.id}
}

export function makeID(factory: IDFactory): ID {
  const id = ID(factory.id)
  factory.id++
  return id
}
