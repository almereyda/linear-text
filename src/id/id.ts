export type ID = number & {[brand]: void}
declare const brand: unique symbol

export function ID(id: number): ID {
  return id as ID
}
