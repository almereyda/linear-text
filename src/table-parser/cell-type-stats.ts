import type {CellType} from './cell-type'

/** Cell type statistics. Missing cells are uncounted. */
export type CellTypeStats = Record<CellType, number>

export function CellTypeStats(): CellTypeStats {
  return {text: 0, number: 0, bool: 0, blank: 0, empty: 0}
}
