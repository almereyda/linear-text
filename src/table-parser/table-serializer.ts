import {Line} from '../line/line'
import type {Group} from '../table/group'
import type {LineBreak} from '../table/line-break'
import type {Table} from '../table/table'

/** Serialize a table to a newline-delimited value. */
export function serializeTable(table: Readonly<Table>): string {
  const result = table.groups
    .map(group => serializeGroup(group, table.lineBreak))
    .join(table.lineBreak) // the delimiter of every group is a line break but that's in the group
  const trailingLineBreak =
    table.groups[table.groups.length - 1]?.lines.length === 0 &&
    table.groups[table.groups.length - 2]?.lines.length === 0
  return trailingLineBreak
    ? result
    : result.substring(0, result.length - table.lineBreak.length)
}

export function serializeGroup(
  group: Readonly<Group>,
  lineBreak: LineBreak
): string {
  return group.lines
    .filter(line => !Line.isEmpty(line))
    .map(({text}) => `${text}${lineBreak}`)
    .join('') // the delimiter of every line is a line break // Every group contains a line break
}
