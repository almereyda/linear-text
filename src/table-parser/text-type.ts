import {isNumeric} from '../utils/string-util'

/** String-encoded data type of the text value. */
export type TextType =
  | 'uri-data'
  | 'uri-http'
  | 'uri-file'
  /** Nonempty string. */
  | 'text'
  /** Finite number. */
  | 'number'
  /** True or false. */
  | 'bool'
  /** Whitespace-only string. */
  | 'blank'
  /** No value. */
  | 'empty'

/** Infer the string-encoded data type of the text value. */
export function parseTextType(text: string): TextType {
  if (text === '') return 'empty'
  if (/^\s+$/.test(text)) return 'blank'
  const trimmed = text.trim()
  if (/^(true|false)$/i.test(trimmed)) return 'bool'
  if (isNumeric(trimmed)) return 'number'
  if (/^file:\/\//.test(text)) return 'uri-file'
  if (/^https?:\/\//.test(text)) return 'uri-http'
  return 'text'
}
