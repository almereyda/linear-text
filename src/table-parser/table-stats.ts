import type {CellType} from './cell-type'
import type {ColumnMap} from '../table/column-map'
import type {Row} from '../table/row'

import {CellTypeStats} from './cell-type-stats'
import {parseCellType} from './cell-type'

export type TableStats = Readonly<{
  /**
   * Column index to CellTypeStats. The maximum number of columns in any row of
   * the table is the length of this array.
   */
  columns: CellTypeStats[]
  /** Aggregated stats for all columns. */
  total: CellTypeStats
}>

export function TableStats(): TableStats {
  return Object.freeze({columns: [], total: CellTypeStats()})
}

/** Tally rows. */
TableStats.sample = (
  stats: TableStats,
  ...rows: readonly Readonly<Row>[]
): void => {
  rows.forEach(row => sampleRow(stats, row))
}

function sampleRow({columns, total}: TableStats, row: Readonly<Row>): void {
  row.forEach((cell, index) => {
    const type = parseCellType(cell)
    // [todo]: ??=.
    if (columns[index] == null) columns[index] = CellTypeStats()
    columns[index]![type]++
    total[type]++
  })
}

TableStats.inferMap = (stats: TableStats): ColumnMap => {
  const maxColumns = stats.columns.length
  return {
    text: findMaxCellTypeIndex(stats, 'text') ?? maxColumns
  }
}

function findMaxCellTypeIndex(
  stats: TableStats,
  type: CellType
): number | undefined {
  if (stats.total[type] === 0) return
  return stats.columns
    .map(columns => columns[type])
    .reduce(
      (maxCountIndex, count, index, types) =>
        count > types[maxCountIndex]! ? index : maxCountIndex,
      0
    )
}
