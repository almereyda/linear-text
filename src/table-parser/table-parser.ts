import type {ColumnMap} from '../table/column-map'
import type {IDFactory} from '../id-factory/id-factory'
import type {Row} from '../table/row'
import type {Table} from '../table/table'

import {TableHeader, tableMetaDefaults} from '../table/table-meta'
import Papa from 'papaparse'
import {Line} from '../line/line'
import {TableStats} from './table-stats'

export function parseTable(
  factory: IDFactory,
  input: File | string
): Promise<Table> {
  return new Promise((resolve, reject) =>
    Papa.parse<Row>(input, {
      complete({errors, meta, data}) {
        const delimiterErrorIndex = errors.findIndex(
          ({code}) => code === 'UndetectableDelimiter'
        )
        if (delimiterErrorIndex > -1) {
          errors.splice(delimiterErrorIndex, 1)
          meta.delimiter = tableMetaDefaults.delimiter
        }
        if (errors.length > 0) return reject(errors)
        return resolve(parse(factory, meta, data))
      }
    })
  )
}

function parse(
  factory: IDFactory,
  parseMeta: Papa.ParseMeta,
  data: Row[]
): Table {
  const {header, rows} = parseTableHeaderAndRows(data)
  let columnMap: ColumnMap = {}
  if (header == null) {
    // No header. Use a heuristic to guess the best mapping from the model
    // columns to the actual row columns.
    const stats = TableStats()
    TableStats.sample(stats, ...rows)
    columnMap = TableStats.inferMap(stats)
  } else {
    // Use the header to map from the model columns to the actual row columns.
    columnMap = {text: findHeaderTextIndex(header)}
  }

  return {
    meta: {
      header,
      columnMap,
      delimiter: parseMeta.delimiter,
      newline: parseMeta.linebreak
    },
    lines: rows.map(row => parseLine(factory, row, columnMap))
  }
}

function parseTableHeaderAndRows(data: readonly Row[]): {
  header: TableHeader | undefined
  rows: readonly Row[]
} {
  if (data[0] == null || !isHeader(data[0]))
    return {header: undefined, rows: data}
  const [header, ...rows] = data
  return {header, rows}
}

function parseLine(factory: IDFactory, row: Row, columnMap: ColumnMap): Line {
  const text = columnMap.text == null ? undefined : row[columnMap.text]
  return Line.fromRow(factory, row, text)
}

function isHeader(row: Readonly<Row>): boolean {
  return findHeaderTextIndex(row) != null
}

function findHeaderTextIndex(row: Readonly<Row>): number | undefined {
  // Find first index of a text column.
  const index = row.findIndex(cell => /^\s*text\s*$/i.test(cell))
  return index === -1 ? undefined : index
}
