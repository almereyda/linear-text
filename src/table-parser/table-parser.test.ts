import type {ReadonlyLine} from '../line/line'
import type {Table} from '../table/table'

import {ID} from '../id/id'
import {IDFactory} from '../id/id-factory'
import {parseTable} from './table-parser'
import {readFileSync} from 'fs'

let factory: IDFactory = IDFactory()
beforeEach(() => (factory = IDFactory()))

test.each(['\t'])('delimiter: "%s"', async delimiter => {
  const row = ['a', 'b', 'c'].join(delimiter)
  const expected: Table = {
    meta: {header: undefined, columnMap: {text: 0}, delimiter, newline: '\n'},
    lines: [{id: ID(1), state: 'note', text: 'a', row: ['a', 'b', 'c']}]
  }
  expect(await parseTable(factory, row)).toStrictEqual(expected)
})

test.each([
  ['linefeed', '\n'],
  ['carriage return and linefeed', '\r\n']
])('newline: %s', async (_, newline) => {
  const input = `text${newline}abc`
  const expected: Table = {
    meta: {header: ['text'], columnMap: {text: 0}, delimiter: '\t', newline},
    lines: [{id: ID(1), state: 'note', text: 'abc', row: ['abc']}]
  }
  expect(await parseTable(factory, input)).toStrictEqual(expected)
})

test('header padding is ignored and preserved', async () => {
  const input = ' a \t text \t c \n 1 \t 2 \t 3 '
  const expected: Table = {
    meta: {
      header: [' a ', ' text ', ' c '],
      columnMap: {text: 1},
      delimiter: '\t',
      newline: '\n'
    },
    lines: [{id: ID(1), state: 'note', text: ' 2 ', row: [' 1 ', ' 2 ', ' 3 ']}]
  }
  expect(await parseTable(factory, input)).toStrictEqual(expected)
})

test('extra empty columns are ignored and preserved', async () => {
  const input = [
    'a							b							text										d',
    '1							2							3										4'
  ].join('\n') // prettier-ignore
  const expected: Table = {
    meta: {
      header: [
        'a', '', '', '', '', '', '',
        'b', '', '', '', '', '', '',
        'text', '', '', '', '', '', '', '', '', '',
        'd'
      ], // prettier-ignore
      columnMap: {text: 14},
      delimiter: '\t',
      newline: '\n'
    },
    lines: [
      {
        id: ID(1),
        state: 'note',
        text: '3',
        row: [
          '1', '', '', '', '', '', '',
          '2', '', '', '', '', '', '',
          '3', '', '', '', '', '', '', '', '', '',
          '4'
        ] // prettier-ignore
      }
    ]
  }
  expect(await parseTable(factory, input)).toStrictEqual(expected)
})

test('header capitalization is ignored and preserved', async () => {
  const input = 'a\tText\tc\n1\t2\t3'
  const expected: Table = {
    meta: {
      header: ['a', 'Text', 'c'],
      columnMap: {text: 1},
      delimiter: '\t',
      newline: '\n'
    },
    lines: [{id: ID(1), state: 'note', text: '2', row: ['1', '2', '3']}]
  }
  expect(await parseTable(factory, input)).toStrictEqual(expected)
})

test.each(<const>[
  ['empty file', '', []],
  [
    'empty file with trailing newline',
    '\n',
    [
      {id: ID(1), state: 'divider', text: undefined, row: ['']},
      {id: ID(2), state: 'divider', text: undefined, row: ['']}
    ]
  ],
  [
    'empty file with header',
    'text\n',
    [{id: ID(1), state: 'divider', text: '', row: ['']}]
  ],
  [
    'nonempty file trailing newline',
    'a\n',
    [
      {id: ID(1), state: 'note', text: 'a', row: ['a']},
      {id: ID(2), state: 'divider', text: '', row: ['']}
    ]
  ],
  [
    'nonempty file with header and trailing newline',
    'text\na\n',
    [
      {id: ID(1), state: <const>'note', text: 'a', row: ['a']},
      {id: ID(2), state: <const>'divider', text: '', row: ['']}
    ]
  ],
  [
    'nonempty file without trailing newline',
    'a\nb',
    [
      {id: ID(1), state: <const>'note', text: 'a', row: ['a']},
      {id: ID(2), state: <const>'note', text: 'b', row: ['b']}
    ]
  ]
])(
  'trailing newline is an empty Line: %s',
  async (_, input, expectedLines: readonly ReadonlyLine[]) => {
    expect((await parseTable(factory, input)).lines).toStrictEqual(
      expectedLines
    )
  }
)

test.each(<const>[
  [
    'missing leading cell',
    '\t2\t3',
    {id: ID(1), state: 'note', text: '2', row: ['', '2', '3']}
  ],
  [
    'missing middling cell',
    '1\t\t3',
    {id: ID(1), state: 'divider', text: '', row: ['1', '', '3']}
  ],
  [
    'missing trailing cell',
    '1\t2',
    {id: ID(1), state: 'note', text: '2', row: ['1', '2']}
  ],
  [
    'missing trailing cell and trailing comma',
    '1\t2\t',
    {id: ID(1), state: 'note', text: '2', row: ['1', '2', '']}
  ],
  [
    'missing row',
    '',
    {id: ID(1), state: 'divider', text: undefined, row: ['']}
  ],
  [
    'extra cell',
    '1\t2\t3\t4',
    {id: ID(1), state: 'note', text: '2', row: ['1', '2', '3', '4']}
  ]
])(
  'missing cells and inconsistencies: %s',
  async (_, row, expectedLine: ReadonlyLine) => {
    const header = 'a\ttext\tc\n'
    const input = header + row
    const expected = {
      meta: {
        header: ['a', 'text', 'c'],
        columnMap: {text: 1},
        delimiter: '\t',
        newline: '\n'
      },
      lines: [expectedLine]
    }
    expect(await parseTable(factory, input)).toStrictEqual(expected)
  }
)

test.each([
  'alphabet.test.txt',
  'colors.test.txt',
  'digits.test.txt',
  'empty.test.txt',
  'games.test.txt',
  'groceries.test.txt',
  'perf-1k.txt',
  'perf-10k.txt'
])('integration %s', async filename => {
  const absoluteFilename = `${__dirname}/../table-test-data/${filename}`
  const input = readFileSync(absoluteFilename).toString()
  expect(await parseTable(factory, input)).toMatchSnapshot()
})
