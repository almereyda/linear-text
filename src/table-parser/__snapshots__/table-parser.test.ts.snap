// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration games.test.tab 1`] = `
Object {
  "lines": Array [
    Object {
      "id": 1,
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "id": 2,
      "invalidated": false,
      "row": Array [
        "Nintendo Entertainment System",
        "1985",
        "https://wikipedia.org/wiki/Nintendo_Entertainment_System",
      ],
      "text": "Nintendo Entertainment System",
    },
    Object {
      "id": 3,
      "invalidated": false,
      "row": Array [
        "- Mega Man 6",
        "1994",
        "https://wikipedia.org/wiki/Mega_Man_6",
      ],
      "text": "- Mega Man 6",
    },
    Object {
      "id": 4,
      "invalidated": false,
      "row": Array [
        "- Solar Jetman: Hunt for the Golden Warpship",
        "1990",
        "https://wikipedia.org/wiki/Solar_Jetman",
      ],
      "text": "- Solar Jetman: Hunt for the Golden Warpship",
    },
    Object {
      "id": 5,
      "invalidated": false,
      "row": Array [
        "- Super Mario Bros. 3",
        "1989",
        "https://wikipedia.org/wiki/Super_Mario_Bros._3",
      ],
      "text": "- Super Mario Bros. 3",
    },
    Object {
      "id": 6,
      "invalidated": false,
      "row": Array [
        "- Teenage Mutant Ninja Turtles III: The Manhattan Project",
        "1992",
        "https://wikipedia.org/wiki/Teenage_Mutant_Ninja_Turtles_III:_The_Manhattan_Project",
      ],
      "text": "- Teenage Mutant Ninja Turtles III: The Manhattan Project",
    },
    Object {
      "id": 7,
      "invalidated": false,
      "row": Array [
        "- The Guardian Legend",
        "1989",
        "https://wikipedia.org/wiki/The_Guardian_Legend",
      ],
      "text": "- The Guardian Legend",
    },
    Object {
      "id": 8,
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "id": 9,
      "invalidated": false,
      "row": Array [
        "PC",
      ],
      "text": "PC",
    },
    Object {
      "id": 10,
      "invalidated": false,
      "row": Array [
        "- Red Dead Redemption 2",
        "2018",
        "https://wikipedia.org/wiki/Red_Dead_Redemption_2",
      ],
      "text": "- Red Dead Redemption 2",
    },
    Object {
      "id": 11,
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "id": 12,
      "invalidated": false,
      "row": Array [
        "PlayStation 3",
        "2006",
        "https://wikipedia.org/wiki/PlayStation_3",
      ],
      "text": "PlayStation 3",
    },
    Object {
      "id": 13,
      "invalidated": false,
      "row": Array [
        "- Ico (The Ico & Shadow of the Colossus Collection)",
        "2001",
        "https://wikipedia.org/wiki/The_Ico_%26_Shadow_of_the_Colossus_Collection",
      ],
      "text": "- Ico (The Ico & Shadow of the Colossus Collection)",
    },
    Object {
      "id": 14,
      "invalidated": false,
      "row": Array [
        "- Metal Gear Solid 2: Sons of Liberty (Metal Gear Solid HD Collection)",
        "2001",
        "https://wikipedia.org/wiki/Metal_Gear_Solid_HD_Collection",
      ],
      "text": "- Metal Gear Solid 2: Sons of Liberty (Metal Gear Solid HD Collection)",
    },
    Object {
      "id": 15,
      "invalidated": false,
      "row": Array [
        "- Shadow of the Colossus (The Ico & Shadow of the Colossus Collection)",
        "2005",
        "https://wikipedia.org/wiki/The_Ico_%26_Shadow_of_the_Colossus_Collection",
      ],
      "text": "- Shadow of the Colossus (The Ico & Shadow of the Colossus Collection)",
    },
    Object {
      "id": 16,
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "id": 17,
      "invalidated": false,
      "row": Array [
        "PlayStation 4",
        "2013",
        "https://wikipedia.org/wiki/PlayStation_4",
      ],
      "text": "PlayStation 4",
    },
    Object {
      "id": 18,
      "invalidated": false,
      "row": Array [
        "- The Last of Us Remastered",
        "2013",
        "https://wikipedia.org/wiki/The_Last_of_Us#The_Last_of_Us_Remastered",
      ],
      "text": "- The Last of Us Remastered",
    },
    Object {
      "id": 19,
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "id": 20,
      "invalidated": false,
      "row": Array [
        "Super Nintendo",
        "1991",
        "https://wikipedia.org/wiki/Super_Nintendo_Entertainment_System",
      ],
      "text": "Super Nintendo",
    },
    Object {
      "id": 21,
      "invalidated": false,
      "row": Array [
        "- EarthBound",
        "1995",
        "https://wikipedia.org/wiki/EarthBound",
      ],
      "text": "- EarthBound",
    },
    Object {
      "id": 22,
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "id": 23,
      "invalidated": false,
      "row": Array [
        "Switch",
      ],
      "text": "Switch",
    },
    Object {
      "id": 24,
      "invalidated": false,
      "row": Array [
        "- The Legend of Zelda: Breath of the Wild",
        "2017",
        "https://wikipedia.org/wiki/The_Legend_of_Zelda:_Breath_of_the_Wild",
      ],
      "text": "- The Legend of Zelda: Breath of the Wild",
    },
    Object {
      "id": 25,
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
  ],
  "meta": Object {
    "columnMap": Object {
      "text": 0,
    },
    "delimiter": "	",
    "header": Array [
      "text",
      "year",
      "wiki",
    ],
    "newline": "
",
  },
}
`;

exports[`integration groceries.test.tab 1`] = `
Object {
  "lines": Array [
    Object {
      "id": 1,
      "invalidated": false,
      "row": Array [
        "Produce",
      ],
      "text": "Produce",
    },
    Object {
      "id": 2,
      "invalidated": false,
      "row": Array [
        "[] Cucumber",
      ],
      "text": "[] Cucumber",
    },
    Object {
      "id": 3,
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "id": 4,
      "invalidated": false,
      "row": Array [
        "Snacks",
      ],
      "text": "Snacks",
    },
    Object {
      "id": 5,
      "invalidated": false,
      "row": Array [
        "[x] Pretzels",
      ],
      "text": "[x] Pretzels",
    },
    Object {
      "id": 6,
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "id": 7,
      "invalidated": false,
      "row": Array [
        "Frozen",
      ],
      "text": "Frozen",
    },
    Object {
      "id": 8,
      "invalidated": false,
      "row": Array [
        "[x] Ice Cream",
      ],
      "text": "[x] Ice Cream",
    },
    Object {
      "id": 9,
      "invalidated": false,
      "row": Array [
        "[x] Pizza",
      ],
      "text": "[x] Pizza",
    },
    Object {
      "id": 10,
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
  ],
  "meta": Object {
    "columnMap": Object {
      "text": 0,
    },
    "delimiter": "	",
    "header": undefined,
    "newline": "
",
  },
}
`;
